sgx/src/enclave/enclave.rs

```diff
@@ -141,41 +115,63 @@ impl Thread {
         Some(Self { enc, tcs, fnc })
     }
 
-    /// Issues `EENTER` instruction to the enclave.
+    /// Enter an enclave.
     ///
-    /// TODO add more comprehensive docs
+    /// This function enters an enclave using `Leaf` and provides the
+    /// specified argument to the enclave in the `rdx` register. On success,
+    /// the value of the `rdx` register at enclave exit is returned. There
+    /// are two failure cases. If the enclave performs an asynchronous exit
+    /// (AEX), the details of the exception are returned. Otherwise, an
+    /// unknown error has occurred.
     #[inline(always)]
     pub fn enter(
         &self,
-        rdi: usize,
-        rsi: usize,
-        rdx: usize,
-        leaf: Leaf,
-        r8: usize,
-        r9: usize,
-    ) -> Result<(), Option<ExceptionInfo>> {
+        how: Leaf,
+        arg: impl Into<Register<usize>>,
+    ) -> Result<Register<usize>, Option<ExceptionInfo>> {
         const FAULT: i32 = -libc::EFAULT;
         const EEXIT: i32 = 0;
 
         let mut exc = MaybeUninit::<ExceptionInfo>::uninit();
+        let mut rdx: usize = arg.into().into();
+        let rax: i32;
+
+        unsafe {
+            asm!(
+                "push rbp",
+                "mov  rbp, rsp",
+                "and  rsp, ~0xf",
```

Comment:
The bitwise arithmetic cuts off the least significant digit of the hex value of `rsp`, but why?

Reply:
Is this just creating space on the stack?

Reply:
To ensure stack alignment to 16 bytes.

With inline assembly, the value of `rsp` is guaranteed to be valid but everything else about it is undefined. So this code works like this:

1. Save `rbp` on the stack and save `rsp` in `rbp`.
2. Align the stack to 16 bytes.
3. Pad the stack.
4. Push function arguments to the stack.
5. Call the function (stack **MUST** be 16-bytes aligned here).
6. Restore `rsp` and `rbp` to their original values.

Reply:
Thanks @npmccallum